app/handlers/repos/hooks.go

```diff
@@ -16,65 +16,92 @@ import (
 	gh "github.com/google/go-github/github"
 )
 
-func receiveGithubWebhook(ctx context.C) error {
+func getWebhookPayload(ctx context.C) (*gh.PullRequestEvent, error) {
 	var payload gh.PullRequestEvent
 	if err := json.NewDecoder(ctx.R.Body).Decode(&payload); err != nil {
-		return herrors.New400Errorf("invalid payload json: %s", err)
+		return nil, herrors.New400Errorf("invalid payload json: %s", err)
 	}
 
 	if payload.PullRequest == nil {
 		ctx.L.Infof("Got webhook without PR")
-		return nil
+		return nil, nil
 	}
 
 	if payload.GetAction() != "opened" && payload.GetAction() != "synchronize" {
 		ctx.L.Infof("Got webhook with action %s, skip it", payload.GetAction())
-		return nil
+		return nil, nil
 	}
 
-	prNumber := payload.PullRequest.GetNumber()
-	if prNumber == 0 {
-		return fmt.Errorf("got zero pull request number: %+v", payload.PullRequest)
+	if payload.PullRequest.GetNumber() == 0 {
+		return nil, herrors.New400Errorf("got zero pull request number: %+v", payload.PullRequest)
 	}
 
+	return &payload, nil
+}
+
+func fetchGithubRepo(ctx context.C) (*models.GithubRepo, error) {
 	var gr models.GithubRepo
 	hookID := ctx.URLVar("hookID")
 	err := models.NewGithubRepoQuerySet(db.Get(&ctx)).
 		HookIDEq(hookID).
 		One(&gr)
 	if err != nil {
-		return fmt.Errorf("can't get github repo with hook id %q: %s", hookID, err)
+		return nil, fmt.Errorf("can't get github repo with hook id %q: %s", hookID, err)
 	}
 
 	if gr.Name != fmt.Sprintf("%s/%s", ctx.URLVar("owner"), ctx.URLVar("name")) {
-		return herrors.New400Errorf("invalid reponame: expected %q", gr.Name)
+		return nil, herrors.New400Errorf("invalid reponame: expected %q", gr.Name)
 	}
 
-	var ga models.GithubAuth
-	err = models.NewGithubAuthQuerySet(db.Get(&ctx)).
-		UserIDEq(gr.UserID).
-		One(&ga)
-	if err != nil {
-		return herrors.New(err, "can't get github auth for user %d", gr.UserID)
-	}
+	return &gr, nil
+}
 
+func createAnalysis(ctx context.C, pr *gh.PullRequest, gr *models.GithubRepo) (*models.GithubAnalysis, error) {
 	guid := ctx.R.Header.Get("X-GitHub-Delivery")
 	if guid == "" {
-		return herrors.New400Errorf("delivery without GUID")
+		return nil, herrors.New400Errorf("delivery without GUID")
 	}
+
 	analysis := models.GithubAnalysis{
 		GithubRepoID:            gr.ID,
-		GithubPullRequestNumber: prNumber,
+		GithubPullRequestNumber: pr.GetNumber(),
```

Comment:
invalid operation: pr (variable of type *invalid type) has no field or method GetNumber